src/Prefetch.cpp

```diff
@@ -0,0 +1,348 @@
+        Expr prefetch_buf = Call::make(type_of<struct buffer_t *>(), Call::create_buffer_t,
+                              args, Call::Intrinsic);
+        plets.push_back(make_pair(varname_prefetch_buf, prefetch_buf));
+
+        // Create the prefetch call
+        Expr num_elem = stride_var[dims-1] * extent_var[dims-1];
+        vector<Expr> args_prefetch(4);
+        args_prefetch[0] = dims;
+        args_prefetch[1] = elem_size_bytes;
+        args_prefetch[2] = num_elem;
+        args_prefetch[3] = var_prefetch_buf;
+        Stmt stmt_prefetch = Evaluate::make(Call::make(Int(32), Call::prefetch_buffer_t,
+                              args_prefetch, Call::Intrinsic));
+        // TODO: Opt: Keep running sum of bytes prefetched on this sequence?
+        // TODO: Opt: Keep running sum of number of prefetch instructions issued
+        // TODO       on this sequence? (to not exceed MAX_PREFETCH)
+        // TODO: Opt: Generate more control code for prefetch_buffer_t in Prefetch.cpp?
+        // TODO       Passing box info through a buffer_t results in ~30 additional stores/loads
+
+        pstmts.push_back(stmt_prefetch);
+    }
+
+    // Inject the generated prefetch code
+    //
+    // Note: plets/pstmts may be modified or partially emptied during injection
+    //
+    void inject_prefetch_stmts(const For *op, const Prefetch &p,
+                vector<pair<string, Expr>> &plets, vector<Stmt> &pstmts, Stmt &body)
+    {
+        if (pstmts.size() > 0) {
+            Stmt pbody = pstmts.back();    // Initialize prefetch body
+            pstmts.pop_back();
+            for (size_t i = pstmts.size(); i > 0; i--) {
+                pbody = Block::make(pstmts[i-1], pbody);
+            }
+            for (size_t i = plets.size(); i > 0; i--) {
+                pbody = LetStmt::make(plets[i-1].first, plets[i-1].second, pbody);
+            }
+
+            // No guard needed, address range checked in prefetch runtime
+            body = Block::make({pbody, body});
+            debug(dbg_prefetch4) << pbody << "\n";
+        }
+    }
+
+    void visit(const Let *op) {
+        Interval in = bounds_of_expr_in_scope(op->value, scope);
+        scope.push(op->name, in);
+        debug(dbg_prefetch5) << "Let scope.push(" << op->name << ")\n";
+        IRMutator::visit(op);
+        debug(dbg_prefetch5) << "Let scope.pop(" << op->name << ")\n";
+        scope.pop(op->name);
+    }
+
+    void visit(const LetStmt *op) {
+        Interval in = bounds_of_expr_in_scope(op->value, scope);
+        scope.push(op->name, in);
+        debug(dbg_prefetch5) << "LetStmt scope.push(" << op->name << ")\n";
+        IRMutator::visit(op);
+        debug(dbg_prefetch5) << "LetStmt scope.pop(" << op->name << ")\n";
+        scope.pop(op->name);
+    }
+
+    void visit(const Realize *op) {
+        rscope.push(op->name, 1);
+        debug(dbg_prefetch5) << "Realize push(" << op->name << ")\n";
+        IRMutator::visit(op);
+    }
```

Comment:
That's working.  Compared to using the realize scope, the has_realization visitor is picking up one more prefetch in HelloHexagon (input_bounded) and two more in camera_pipe (curve & matrix) but we're dropping that prefetch directive anyway:

Runtimes before and after this change are extremely close to each other.

% env HL_DEBUG_PREFETCH=1 test-hellohexagon
...
prefetch:
 blur prefetch(y, 2)
  Info: not prefetching realized blur_y
  prefetch #0: input (input.elem_size, dims:3)
  prefetch #1: input_bounded (uint8, dims:3)


% env HL_DEBUG_PREFETCH=1 test-camera-android-128
...
prefetch:
 denoised prefetch(v1, 2)
  prefetch #0: input (input.elem_size, dims:2)
 processed prefetch(yo, 2)
  Info: not prefetching realized corrected
  prefetch #1: curve (uint8, dims:1)
  Info: not prefetching realized deinterleaved
  Info: not prefetching realized demosaiced
  Info: not prefetching realized denoised
  Info: not prefetching realized g_b
  Info: not prefetching realized g_r
  prefetch #2: input (input.elem_size, dims:2)
  prefetch #3: matrix (int16, dims:2)